components/vr/webvr_thread.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+#[allow(unsafe_code)]
+unsafe impl Send for WebVRCompositor {}
+
+pub type WebVRCompositorSender = Sender<Option<WebVRCompositor>>;
+
+impl WebVRCompositorHandler {
+    pub fn new() -> (Box<WebVRCompositorHandler>, WebVRCompositorSender) {
+        let (sender, receiver) = mpsc::channel();
+        let instance = Box::new(WebVRCompositorHandler {
+            compositors: HashMap::new(),
+            webvr_thread_receiver: receiver,
+            webvr_thread_sender: None
+        });
+
+        (instance, sender)
+    }
+}
+
+impl webrender_traits::VRCompositorHandler for WebVRCompositorHandler {
+    #[allow(unsafe_code)]
+    fn handle(&mut self, cmd: webrender_traits::VRCompositorCommand, texture_id: Option<u32>) {
+        match cmd {
+            webrender_traits::VRCompositorCommand::Create(compositor_id) => {
+                self.create_compositor(compositor_id);
+            }
+            webrender_traits::VRCompositorCommand::SyncPoses(compositor_id, near, far, sender) => {
+                if let Some(compositor) = self.compositors.get(&compositor_id) {
+                    let pose = unsafe {
+                        (*compositor.0).sync_poses();
+                        (*compositor.0).synced_frame_data(near, far).to_bytes()
+                    };
+                    let _ = sender.send(Ok(pose));
+                } else {
+                    let _ = sender.send(Err(()));
+                }
+            }
+            webrender_traits::VRCompositorCommand::SubmitFrame(compositor_id, left_bounds, right_bounds) => {
+                if let Some(compositor) = self.compositors.get(&compositor_id) {
+                    if let Some(texture_id) = texture_id {
+                        let layer = VRLayer {
+                            texture_id: texture_id,
+                            left_bounds: left_bounds,
+                            right_bounds: right_bounds
+                        };
+                        unsafe {
+                            (*compositor.0).submit_frame(&layer);
+                        }
+                    }
+                }
+            }
+            webrender_traits::VRCompositorCommand::Release(compositor_id) => {
+                self.compositors.remove(&compositor_id);
+            }
+        }
+    }
+}
+
+impl WebVRCompositorHandler {
+    #[allow(unsafe_code)]
+    fn create_compositor(&mut self, device_id: webrender_traits::VRCompositorId) {
+        let sender = match self.webvr_thread_sender {
+            Some(ref s) => s,
+            None => return,
+        };
+
+        sender.send(WebVRMsg::CreateCompositor(device_id)).unwrap();
+        let device = self.webvr_thread_receiver.recv().unwrap();
+
+        match device {
+            Some(device) => {
+                self.compositors.insert(device_id, device);
+            },
+            None => {
+                error!("VRDevice not found when creating a new VRCompositor");
```

Comment:
`VRDisplay` -> `VRDisplay` (or `VR Display`, if want to be consistent everywhere w/ that usage)